### PR TITLE
perf: adapt SGLang LightOp operators for vLLM 0.25.1

### PR DESCRIPTION
**Pull request by alexanderbin123**
## Summary

Port the useful SGLang-DAS LightOp/AITER candidates to vLLM-DAS v0.25.1 through plugin-owned seams, with strict eligibility, official fallback, HCU CI registration, and production-entrypoint benchmarks.

Source identities used for the operator audit:
- `HYGON-AI/sglang-das`: `0457572a14edd55e05220726ff4a8af8b1931771`
- vLLM source: `7b108ad1` / installed vLLM 0.25.1
- operator implementation head: `47952ee`; current PR head `6732a60` only merges the updated `v0.25.1` target branch

Accepted routes:
- LightOp non-hash sqrt-softplus MoE routing
- LightOp Qwen3.5/Qwen3.6 gated RMSNorm
- LightOp `ds_cat` for the exact FlashMLA decode concat seam
- LightOp BF16 W16A16 Marlin MoE (opt-in because it changes weight layout)

## Delivery boundary

This PR intentionally does **not** ship `.agents/**`, operator docs/plans, README edits, or workflow files. The reusable skill and detailed working records are local-only and are not PR deliverables or reproducibility references. The release/rollback contract and final production evidence are recorded directly below; the executable benchmark and tests remain in this diff.

## Runtime controls and rollback

All routes require `VLLM_HCU_USE_CUSTOM_OPS=1` (default: on). Set `VLLM_HCU_USE_CUSTOM_OPS=0` before process start to disable all four routes.

| Route | Leaf default | Per-route rollback | Fallback behavior |
|---|---:|---|---|
| sqrt-softplus | on | `VLLM_HCU_USE_LIGHTOP_SQRTSOFTPLUS_GATE=0` | Official vLLM for disabled, missing LightOp, hash routing, or any ineligible input. |
| Qwen gated RMSNorm | on | `VLLM_HCU_USE_LIGHTOP_QWEN_RMSNORM_GATED=0` | Canonical vLLM Triton path for disabled, missing LightOp, or any ineligible input. |
| MLA decode concat | on | `VLLM_HCU_USE_LIGHTOP_MLA_DECODE_CAT=0`; `VLLM_USE_OPT_CAT=0` is also an opt-out | `torch.cat` for disabled, missing LightOp, or any ineligible layout. |
| W16A16 Marlin MoE | **off** | Keep `VLLM_HCU_USE_LIGHTOP_W16A16_MOE=0` and restart before model load | AITER/Triton selection occurs before packing. After packing, execution failures fail closed; canonical reload/fallback is unsafe and the model process must be recreated. |

Optional dependencies are imported lazily. For an eligible route, execution/ABI failures propagate unless the registered op has a documented safe fallback; unsupported inputs are rejected before mutation.

### sqrt-softplus

Requires CUDA/HCU, contiguous FP32 logits and bias on the same device, non-hash routing, and exactly one `(experts, top-k, tokens)` tuple:

`(256,6,1024)`, `(256,8,256)`, `(256,8,1024)`, `(384,6,1024)`, `(384,8,256)`, `(384,8,1024)`.

All other shapes stay on the official router.

### Qwen gated RMSNorm

Only the exact captured Qwen GDN `RMSNormGated` binding is patched. Inputs, gate and weight must be contiguous BF16 on one HCU; width must be 128 or 256; bias is absent; grouping is full-width; `norm_before_gate=True`; activation is SiLU. Admitted rows are:

`1,2,4,6,8,12,16,24,32,48,64,96,128,144,192,256,384,512,768,1024`.

### MLA decode concat

Requires BF16 FlashMLA decode tensors with widths `512+64`, the validated non-contiguous FlashMLA strides, and one `(tokens, heads)` pair:

- heads 8: tokens 256/512/768
- heads 16: tokens 128/256/512/768
- heads 32: tokens 64/128/256/512/768

### W16A16 Marlin MoE

Default-off and restricted to BF16 `E=256,K=2048,N=512,top-k=8`, standard unmodified SwiGLU, `max_num_tokens` 1 through 16, TP-only, no bias, no EP/EPLB/DP/PCP/SP, and router weights applied inside the expert path. Every reachable M config is checked before packing. Packed tensors carry layout/generation markers; repeated setup is idempotent, stale/partial markers fail closed, and post-pack weight reload raises an explicit recreate-model error.

## Final production-path evidence

Environment: BW1100/gfx938, torch 2.11.0, installed vLLM 0.25.1, LightOp 0.6.0, seed 20260906. Benchmarks use the real production entrypoint, 50 warmups, 200 iterations, seven repeated medians, required allocations/conversions, and synchronization.

### sqrt-softplus: production allowlist 6/6

Live accuracy covered every allowlisted shape with `(renormalize=True, scale=1.5)` and `(False, 1.0)`: IDs matched exactly after ordering; weights matched an independent FP32 expression at `rtol=1e-5, atol=1e-6`; the accuracy file passed 15/15.

| E/top-k/tokens | speedup vs official | max abs weight error |
|---|---:|---:|
| 256/6/1024 | 21.45% | 1.1921e-7 |
| 256/8/256 | 8.95% | 8.9407e-8 |
| 256/8/1024 | 31.89% | 8.9407e-8 |
| 384/6/1024 | 8.57% | 1.1921e-7 |
| 384/8/256 | 14.15% | 8.9407e-8 |
| 384/8/1024 | 10.93% | 1.1921e-7 |

The old 40/40 number described a wider screening corpus. The final benchmark reports **6 accepted production records + 34 explicit `skipped_shapes`**; only these six have final production-wrapper evidence and are routed.

### MLA decode concat: production allowlist 12/12

Every record was compared through `concat_mla_decode` against `torch.cat` using the real FlashMLA strides. All outputs were exact.

| tokens/heads | speedup | max abs error |
|---|---:|---:|
| 256/8 | 21.11% | 0 |
| 512/8 | 48.29% | 0 |
| 768/8 | 48.41% | 0 |
| 128/16 | 22.51% | 0 |
| 256/16 | 48.14% | 0 |
| 512/16 | 70.98% | 0 |
| 768/16 | 70.00% | 0 |
| 64/32 | 20.88% | 0 |
| 128/32 | 47.85% | 0 |
| 256/32 | 64.86% | 0 |
| 512/32 | 61.40% | 0 |
| 768/32 | 57.28% | 0 |

The old 33/33 number described a wider screening corpus. The final benchmark reports **12 accepted production records + 21 explicit `skipped_shapes`**. Excluded combinations stay on `torch.cat` because final hardening admits only shapes with complete production-entrypoint accuracy and stable >5% wins.

### Qwen gated RMSNorm

The production allowlist is the screened matrix itself: 20 admitted row counts x two widths = 40 shapes, all above the 5% gate. Fourteen routed live-HCU cases matched an independent FP32 expression at `rtol=0.02, atol=0.05`; FP32-weight and missing-export fallback cases also passed. The Qwen3.6-35B row-512 case improved 80.43%.

### W16A16 Marlin MoE

`LightopW16A16Experts.apply` was exercised for every M=1..16. All 16 comparisons passed versus Triton and all 16 versus AITER. Speedup was 31.67%–65.97% vs Triton and 5.90%–69.54% vs AITER; max absolute error was 0.0009765625 and 0.00213623046875 respectively.

## Screened AITER/LightOp candidates not routed

- AITER SiLU: accuracy passed but slower than the current LightOp/vLLM paths.
- AITER `tgemm`: 36/36 exact; 0/36 met the 5% gate.
- AITER BF16 batched GEMM (`WO_A`): 6/6 accuracy vs independent FP32; 0/28 met the 5% gate (2.04x–5.24x baseline latency).
- AITER fused recurrent packed decode and LayerNorm2D: rejected by ABI/accuracy/performance gates.
- EP m-index, RMS+quant, sparse MLA and DSv4 cache/attention families: already covered by plugin-owned routes.
- Generic KV fusion / fused QK-RMS-RoPE-cache: no equivalent vLLM 0.25.1 paged-cache ownership seam.

## Model evidence

- `/models/Qwen3.6-35B-A3B`: feature-off/on each produced and reviewed exactly 32 HumanEval samples; Pass@1 was 1.0 on both. Observed output TPS was 15.09 -> 17.24. Fresh feature-on bytes contained W16A16 and gated-RMSNorm markers; feature-off did not.
- `/models/DeepSeek-V4-Flash-Channel-INT8-w8a8`: feature-off/on each produced and reviewed exactly 32 HumanEval samples; Pass@1 was 1.0 on both and TPS 7.12 -> 7.12. This request stream did not hit the final six sqrt allowlist shapes, so no fresh marker was expected or observed; it is an accuracy/fallback control, not positive route coverage.

Model coverage is supplementary to direct operator accuracy; unsupported model geometry remains on the official path.

## Reproduction and repository verification

```bash
VLLM_PLUGINS= python3 tools/run_patch_tests.py --suite contract \
  --vllm-source /models/zb/vllm_025/vllm
python3 tools/run_patch_tests.py --suite inventory \
  --vllm-source /models/zb/vllm_025/vllm
HIP_VISIBLE_DEVICES=4 python3 tools/run_patch_tests.py --suite accuracy-hcu \
  --vllm-source /models/zb/vllm_025/vllm \
  --target tests/accuracy/test_lightop_sqrtsoftplus_gate.py -- -s
HIP_VISIBLE_DEVICES=4 python3 tools/benchmark_sglang_operator_candidates.py \
  sqrtsoftplus-gate --warmup 50 --iterations 200 --repeats 7 \
  --json-output /tmp/vllm-hcu-sqrtsoftplus-production-accuracy-filtered.json
HIP_VISIBLE_DEVICES=4 python3 tools/benchmark_sglang_operator_candidates.py \
  mla-decode-cat --warmup 50 --iterations 200 --repeats 7 \
  --json-output /tmp/vllm-hcu-mla-decode-cat-production-filtered.json
HIP_VISIBLE_DEVICES=4 python3 tools/benchmark_sglang_operator_candidates.py \
  w16a16-moe --warmup 50 --iterations 200 --repeats 7 \
  --json-output /tmp/vllm-hcu-w16a16-production-apply.json
python3 tools/check_patch_test_coverage.py --json
python3 tools/check_production_boundary.py
python3 -m compileall -q vllm_hcu tests tools
git diff --check
```

Final local results on the operator head:
- portable contract: 1514 passed, 163 deselected
- inventory: 82 passed
- patch coverage: 98 files, 97 adapters, one helper, zero untested/invalid contracts
- production boundary: 259 Python files clean
- independent final review: Critical 0, Important 0

Current GitHub status on `6732a60`: static gate, repository checks, both HCU accuracy jobs, both HCU contract partitions, and integration smoke are green. The aggregate CI gate is red because three EvalScope model jobs attempted to load GSM8K/MMMU while runner offline mode was enabled and those datasets were not cached; artifact tracebacks fail during dataset loading, not in an operator kernel or assertion. Those jobs still require a rerun after CI data availability is restored.

### CODE REVIEW

**Comment by wics1224** [comment]
[P2] 最终代码的默认启用行为、支持范围和 PR 验证证据已经不同步
最终 head 新增的几条 LightOp 路由中，以下三个开关默认开启：
- VLLM_HCU_USE_LIGHTOP_SQRTSOFTPLUS_GATE
- VLLM_HCU_USE_LIGHTOP_QWEN_RMSNORM_GATED
- VLLM_HCU_USE_LIGHTOP_MLA_DECODE_CAT
W16A16 Marlin MoE 则默认关闭。具体默认值可见 [envs.py 的开关定义](https://github.com/HYGON-AI/vllm-plugin-das/pull/73/files)。
但最后一笔 hardening 提交同时删除了：
- README 中这些开关的默认值、禁用/回滚方法和支持范围说明；
- docs/operator_adaptation_audit_v0251.md；
- docs/sglang_operator_adaptation_validation_v0251.md；
- PR 描述所称的可复用 operator-porting skill 及 validation checklist。
这些删除可以在 [最终提交 47952ee](https://github.com/HYGON-AI/vllm-plugin-das/commit/47952ee) 中直接看到；[当前最终 diff](https://github.com/HYGON-AI/vllm-plugin-das/pull/73/files) 也已经只剩 Python/YAML 文件。与此同时，[PR 描述](https://github.com/HYGON-AI/vllm-plugin-das/pull/73) 仍然把上述 skill、audit/validation 文档和其中记录的复现命令作为本 PR 的交付物及性能证据来源。
性能覆盖数字也没有跟最终生产路由建立对应关系：
- PR 描述仍写 sqrt-softplus 有 40/40 benchmark shapes faster；
- 但最终 [sqrt-softplus 路由 allowlist](https://github.com/HYGON-AI/vllm-plugin-das/pull/73/files) 只允许 6 个精确 shape；
- PR 描述仍写 MLA 33/33 benchmark shapes faster；
- 但最终 [MLA decode-cat allowlist](https://github.com/HYGON-AI/vllm-plugin-das/pull/73/files) 只允许 12 个 (tokens, heads) 组合。
这不一定说明 40/40 或 33/33 的 benchmark 结果本身错误——它们可能来自更宽的测试网格——但现在无法从最终 MR 判断：
1. 最终允许进入生产路径的 6/12 个 shape 是否全部包含在报告中；
2. 每个最终支持 shape 对应的正确性误差和性能数据是什么；
3. 为什么其余 benchmark shape 在最终提交里被排除；
4. 线上出现回归时应通过哪个环境变量禁用哪条路径；
5. W16A16 packed-weight 生命周期及其限制是否仍与被删除文档中的版本一致。
因为三个新路径是默认开启的，这不只是文档整洁度问题，而是发布和回滚契约缺失。
建议修改：
- 至少恢复 README 中的环境变量说明：默认值、支持范围、fallback 条件以及逐算子 rollback 方法；
- 恢复或重写 validation 文档，记录基于最终 47952ee 的命令、软件版本、容差和结果摘要；
- 把 40/40、33/33 拆分为“benchmark corpus”和“production routing allowlist”，明确 6/12 个最终路由 shape 与原始结果的映射；
- 如果文档和 skill 是有意从本 PR 移除，则同步修改 PR 描述，不能继续把它们列为交付物或把已删除的 validation 文档当作复现依据。

**Comment by alexanderbin123** [comment]
已按当前 PR head `6732a60` 同步发布/回滚契约，处理方式遵守本任务的交付边界：不恢复 `.agents/**`、docs、README 或 workflow 改动，详细工作记录继续只保存在本地 skill，不再把它们列为 PR 交付物或复现依据。

PR 描述现已补齐：
- `VLLM_HCU_USE_CUSTOM_OPS` 总控，以及 4 个 leaf 的默认值、逐算子禁用方式和 fallback 边界；
- sqrt-softplus 的 6 个精确生产 shape、MLA 的 12 个精确 `(tokens, heads)` 组合、Qwen gated RMSNorm 的完整 rows/widths 约束；
- W16A16 默认关闭、pre-pack 选择、packed marker/generation、post-pack fail-closed、必须重建模型进程的生命周期；
- sqrt 6/6 和 MLA 12/12 的逐 shape speedup/误差表；原 40/40、33/33 已明确标为历史 screening corpus，不再作为最终生产路由证据；
- 基于最终 operator head `47952ee`（当前 `6732a60` 仅合入目标分支）的环境、容差、命令、仓库验证及 HumanEval-32 覆盖边界。

复核结果：默认值/allowlist 报告契约测试 `24 passed`；skill validation 通过且 broken links 为 0；相对目标 base，`.agents`、docs、README 和 `.github/workflows` 仍均为 0 diff。

当前 static gate、仓库检查、两组 HCU accuracy、HCU contract 和 integration smoke 均通过。`ci-gate` 仍红的直接原因是三个 EvalScope job 在 runner offline mode 下缺少 GSM8K/MMMU 缓存，artifact traceback 停在 dataset loading，不是算子 kernel/assertion；待 CI 数据可用后需要重跑。

**Review by wics1224 — APPROVED** [review]

**Review by wics1224 — APPROVED** [review]
